feat(semantics): emit type casts after doing arg type checks

adnrs96 opened the pull request:
**- What I did**
In this commit we start emitting type casts for the arguments in
function calls, mutations or services.

**- How I did it**
We were already doing the checks for arg compatibility and had
implicit cast method ready as well. In this commit we just refactor
the implicit_cast method a bit to extract out a type_cast_entity
function which works on single entity and add type casts to its AST
by adding as operator to its tree.

**- How to verify it**
As a result, now calling a function which accpets a float parameter
with an int parameter will specifically emit a type cast
(as operator) in the AST.

Closes: #993.

adnrs96 commented on storyscript/compiler/semantics/ExpressionResolver.py:
```diff
@@ -154,6 +154,32 @@ def nary_expression(self, tree, op, values):
+
     def implicit_cast(self, tree, val, values):
         """
         Creates an AST with the cast.
```
I am pretty sure this comment is outdated (even before this PR) since I don't see anything specific to `pow_expression` happening in this function. Thoughts?

adnrs96 replied:
Mentioning you here @wilzbach since I believe you might have missed this one :)

wilzbach replied:
Yes, sorry about overseeing that. Will update this directly on your PR.

adnrs96 commented on storyscript/compiler/semantics/functions/Function.py:
```diff
@@ -48,16 +51,26 @@ def _check_arg_types(self, tree, args):
             target = self._args[k].type()
             t = args[k].type()
-            tree.expect(target.can_be_assigned(t) or t == AnyType.instance(),
+            is_type_cast_feasible = target.can_be_assigned(t)
```
not the best variable name since result of `target.can_be_assigned(t)` is something in `None, datatype (like int, float, string), True, False`.
Suggestions?

wilzbach replied:
`type_cast_result` ?

wilzbach commented on storyscript/compiler/semantics/ExpressionResolver.py:
```diff
@@ -154,6 +154,32 @@ def nary_expression(self, tree, op, values):
+        as_operator
+        Note: This does not perform any checks around feasibility of performing
+        the type cast operation and depends upon the caller to have had
+        performed these checks before making the call.
```
Add a `Args` section too

wilzbach commented on storyscript/compiler/semantics/ExpressionResolver.py:
```diff
@@ -154,6 +154,32 @@ def nary_expression(self, tree, op, values):
+        """
+        Type casts the given expr_node with an entity to the target_type.
+        This is done by creating a new AST for the expr_node with an
+        as_operator
```
```suggestion
        `as_operator`.
```

wilzbach commented on storyscript/compiler/semantics/ExpressionResolver.py:
```diff
@@ -154,6 +154,32 @@ def nary_expression(self, tree, op, values):
+    @staticmethod
+    def type_cast_entity(expr_node, target_type):
+        """
+        Type casts the given expr_node with an entity to the target_type.
```
This is wrong. You cast expressions, not entities.

wilzbach replied:
While you renamed the function name, the description still says `entity`.

adnrs96 replied:
oh boy, I thought I fixed this. I should have payed better attention. Sorry for that.

wilzbach commented on storyscript/compiler/semantics/functions/Function.py:
```diff
@@ -1,3 +1,6 @@
+from storyscript.compiler.semantics.ExpressionResolver import (
+    SymbolExpressionVisitor
+)
```
No need for parentheses here, simply do `\`

adnrs96 replied:
Matter of preference of each community I guess. Zulip encourages use of the brackets.
I will keep in mind that we use `\` here :)

wilzbach replied:
Well, I don't really care about this, but as the rest of the code uses `\`  for imports, it's a good idea to keep that convention.

wilzbach commented on storyscript/compiler/semantics/functions/Function.py:
```diff
@@ -48,16 +51,26 @@ def _check_arg_types(self, tree, args):
         the respective function argument.
         """
+        i = 0
+        c = tree.children
```
```suggestion
        childs = tree.children
```

wilzbach commented on storyscript/compiler/semantics/ExpressionResolver.py:
```diff
@@ -154,6 +154,32 @@ def nary_expression(self, tree, op, values):
+        performed these checks before making the call.
+        """
+        assert expr_node.data == 'expression'
+        casted_type = SymbolExpressionVisitor.type_to_tree(
```
```suggestion
        cast_type = SymbolExpressionVisitor.type_to_tree(
```

Let's improve this slightly if we're already at it.

wilzbach commented on storyscript/compiler/semantics/functions/Function.py:
```diff
@@ -48,16 +51,26 @@ def _check_arg_types(self, tree, args):
+        i = 0
+        c = tree.children
         for k in args.keys():
+            i += 1  # Tracks argument children of tree
```
use `zip` instead.

adnrs96 replied:
Thanks for pointing out. I was actually doing the `i+=1` thing with something else in mind (which wasn't correct).
Updated to use `zip` and it definitely feels more beautiful.

adnrs96 commented:
Hey @wilzbach 
Thanks for reviewing and the feedback. I have addressed the comments and updated this PR and believe it is ready for a look over and a merge thereafter :)

wilzbach commented on storyscript/compiler/semantics/ExpressionResolver.py:
```diff
@@ -154,6 +154,41 @@ def nary_expression(self, tree, op, values):
+                expression for which this function will generate a new
+                expression node which also contains an `as_operator` to
+                represent the type cast.
+            target_type (Union[str, MapType, ListType]): The type to type cast
```
Could be int, float, boolean etc too

adnrs96 replied:
actually the actual data that target_type receives can only be in `Union[str, MapType, ListType]`
it could receive a `str` whose value can be ('string', 'int', 'float', 'boolean').
This is what I had in mind when I wrote the doc string :)
Thoughts @wilzbach on what to do in such cases?

wilzbach commented:
@adnrs96 FYI: in the interest of time, I added https://github.com/storyscript/storyscript/pull/1039/commits/ce62b62e32544ae2465cad0058858b019058db00 to your PR. This addresses one minor comment plus does a small renaming of `implicit_cast` + `types` in the `nary_expression` method.

adnrs96 commented:
Thanks for reviewing and cleaning up those issues @wilzbach.
Actually the rename effort you did was a much needed thing. I wasn't sure why those were named like that before but yeah it makes a lot of sense for anyone reading the code if they see `type` instead of `value` used in those spots.
